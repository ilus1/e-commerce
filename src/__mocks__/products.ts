export const mockProducts = [
  { id: '1', name: 'Produto 1', price: 10, photo: 'url1', description: 'Descrição 1' },
];

export const mockCartItems = [
  { id: '1', name: 'Produto 1', price: 10, photo: 'url1', description: 'Descrição 1', count: 1 },
];
