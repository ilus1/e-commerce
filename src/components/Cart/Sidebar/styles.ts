import styled from "styled-components"
import { Button } from "@mui/material"

import colors from "../../../constants/colors"


export const Container = styled.div`
  position: absolute;
  top: 0;
  right: 0;
  z-index: 1000;

  display: flex;
  flex-direction: column;
  justify-content: space-between;

  height: 100%;
  width: 486px;
  background-color: ${colors.primary};
  box-shadow: -5px 0px 6px 0px rgba(0, 0, 0, 0.13);
`

export const Header = styled.div`
  display: flex;
  justify-content: space-between;

  padding: 20px 30px;
`

export const Title = styled.h2`
  width: 160px;

  font-size: 27px;
  font-weight: 700;
  color: ${colors.background};
`


export const CloseButton = styled.button`
  display: flex;
  align-items: center;
  justify-content: center;

  font-size: 20px;
  font-weight: 400;

  width: 38px;
  height: 38px;
  margin-top: 25px;
  background-color: ${colors.sidebarButtonBackground};
  color: ${colors.background};

  border: none;
  border-radius: 50%;
  cursor: pointer;
`

export const Cost = styled.div`
  display: flex;
  justify-content: space-between;

  margin: 0px 50px;
  margin-top: auto;
`

export const Total = styled.p`
  display: flex;
  justify-content: center;

  font-size: 28px;
  font-weight: 700;
  color: ${colors.background};

  padding: 20px 0;
`

export const BuyButton = styled(Button)`
  display: inline-flex;
  align-items: center;
  justify-content: center;

  font-size: 28px !important;
  font-weight: 700 !important;

  width: 100%;
  min-height: 97px;
  background-color: ${colors.sidebarButtonBackground} !important;
  color: ${colors.background} !important;

  border-radius: 0 !important;
  cursor: pointer;
`
