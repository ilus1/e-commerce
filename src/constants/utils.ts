export const footerMessage = 'MKS sistemas Â© Todos os direitos reservados'

export const apiURL = 'https://mks-frontend-challenge-04811e8151e6.herokuapp.com/api/v1/products?page=1&rows=8&sortBy=id&orderBy=DESC'

export const mountLoaders = (Loader: JSX.Element, amount: number): JSX.Element[] => {
  const loaderArray = []
  for (let i = 0; i < amount; i++) {
    loaderArray.push(Loader)
  }
  return loaderArray
}
