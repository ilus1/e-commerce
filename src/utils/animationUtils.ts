export const sidebarAnimation = {
  y: {
    type: "spring",
    damping: 25,
    stiffness: 120,
    duration: 0.7
  },
  opacity: {
    duration: 0.3
  },
};
