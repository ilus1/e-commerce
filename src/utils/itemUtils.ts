interface CartItem {
  id: string
  name: string
  price: number
  photo: string
  description: string
  count: number
}

export const removeItemFromCart = (items: CartItem[], id: string) => {
  return items.filter(item => item.id !== id);
}

export const countTotalPrice = (items: CartItem[]) => {
  return items.reduce((acc, item) => acc + item.price * item.count, 0);
}
