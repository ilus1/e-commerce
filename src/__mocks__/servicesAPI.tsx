import { rest } from "msw";
import { setupServer } from "msw/node";

import { mockProducts } from "./products";

const server = setupServer(
  rest.get('*', (_, res, ctx) => res(ctx.status(200), ctx.json(mockProducts)))
);

export default server;
