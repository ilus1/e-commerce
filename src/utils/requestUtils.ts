import { apiURL } from '../constants/utils'

export const fetchProducts = async () => {
  return await fetch(apiURL)
  .then((res) => res.json())
}
