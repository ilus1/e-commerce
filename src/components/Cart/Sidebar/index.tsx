import { motion } from 'framer-motion'

import CartItemList from '../CartItemList'
import {
  Container,
  Header,
  Title,
  CloseButton,
  Cost,
  Total,
  BuyButton,
} from './styles'

import { sidebarAnimation } from '../../../utils/animationUtils'
import { countTotalPrice, removeItemFromCart } from '../../../utils/itemUtils'

interface CartItem {
  id: string
  name: string
  price: number
  photo: string
  description: string
  count: number
}

interface SidebarProps {
  isOpen: boolean
  onClose: () => void
  cartItemList: CartItem[]
  onChangeCartItemList: (cartItemList: CartItem[]) => void
  onRemoveItem: (cartItem: CartItem) => void
  onAddItem: (cartItem: CartItem) => void
}

function Sidebar({ isOpen, onClose, cartItemList, onChangeCartItemList, onAddItem, onRemoveItem }: SidebarProps) {
  const handleRemoveProduct = (cartItem: CartItem) => {
    onChangeCartItemList(removeItemFromCart(cartItemList, cartItem.id))
  }

  return (
    <Container
      as={motion.div}
      initial={{ y: '-100%', opacity: 0 }}
      animate={{ y: isOpen ? 0 : '-100%', opacity: isOpen ? 1 : 0 }}
      exit={{ y: '-100%', opacity: 0 }}
      transition={sidebarAnimation}
    >
      <Header>
        <Title>Carrinho de compras</Title>
        <CloseButton onClick={onClose}>X</CloseButton>
      </Header>
      <CartItemList
        cartItemList={cartItemList}
        onRemoveProduct={handleRemoveProduct}
        onRemoveItem={onRemoveItem}
        onAddItem={onAddItem}
      />
      <Cost>
        <Total>Total:</Total>
        <Total>R${countTotalPrice(cartItemList)}</Total>
      </Cost>
      <BuyButton>Finalizar Compra</BuyButton>
    </Container>
  )
}

export default Sidebar
