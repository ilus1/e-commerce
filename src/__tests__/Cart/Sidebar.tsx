import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';

import Sidebar from '../../components/Cart/Sidebar';

import { mockCartItems } from '../../__mocks__/products';
import { countTotalPrice } from '../../utils/itemUtils';

describe('Componente Sidebar do carrinho de compras', () => {

  it('Testa carregamento correto do componente', () => {
    const tree = renderer
      .create(
        <Sidebar
          onClose={() => {}}
          cartItemList={mockCartItems}
          onChangeCartItemList={() => {}}
          onRemoveItem={() => {}}
          onAddItem={() => {}}
        />
      );

    expect(tree).toMatchSnapshot();
  });

  it('Testa se o botão de fechar chama a função onClose', () => {
    const mockOnClose = jest.fn();
    render(
      <Sidebar
        onClose={mockOnClose}
        cartItemList={[]}
        onChangeCartItemList={() => {}}
        onRemoveItem={() => {}}
        onAddItem={() => {}}
      />
    );

    fireEvent.click(screen.getByText('X'));
    expect(mockOnClose).toHaveBeenCalled();
  });

  it('Testa se o total calculado é exibido corretamente', () => {
    render(
      <Sidebar
        onClose={() => {}}
        cartItemList={mockCartItems}
        onChangeCartItemList={() => {}}
        onRemoveItem={() => {}}
        onAddItem={() => {}}
      />
    );

    const totalPrice = countTotalPrice(mockCartItems);
    expect(screen.getByText(`R$${totalPrice}`)).toBeDefined();
  });

  it('Testa se o item e removido corretamente da listagem', () => {
    render(
      <Sidebar
        onClose={() => {}}
        cartItemList={mockCartItems}
        onChangeCartItemList={() => {}}
        onRemoveItem={() => {}}
        onAddItem={() => {}}
      />
    );

    const removeProductButton = screen.queryAllByText('X')[1];
    fireEvent.click(removeProductButton);
  });
});
