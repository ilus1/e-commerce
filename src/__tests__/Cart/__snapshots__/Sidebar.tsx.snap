// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Componente Sidebar do carrinho de compras Testa carregamento correto do componente 1`] = `
<div
  className="sc-pGaPU eZqOcd"
  style={
    {
      "opacity": 0,
      "transform": "translateY(-100%) translateZ(0)",
    }
  }
>
  <div
    className="sc-jrAGKZ eDyfqU"
  >
    <h2
      className="sc-kEjbdu bAzOSX"
    >
      Carrinho de compras
    </h2>
    <button
      className="sc-iqHXzD jctJuk"
      onClick={[Function]}
    >
      X
    </button>
  </div>
  <ul
    className="sc-fubDmA lihiSQ"
  >
    <li
      className="sc-hKgKIp fFLrTl"
      data-testid="cart-item-component"
      style={
        {
          "opacity": 0,
          "transform": "translateX(40px) translateZ(0)",
        }
      }
    >
      <button
        className="sc-eCstZk bWQkgF"
        onClick={[Function]}
      >
        X
      </button>
      <img
        className="sc-gKsecS hntZsr"
        src="url1"
      />
      <span
        className="sc-jSguLX dDFcMB"
      >
        Produto 1
      </span>
      <div
        className="sc-bdfCDU fhrPBJ"
      >
        <button
          className="sc-dlfmHC jhxSjB"
          onClick={[Function]}
        >
          -
        </button>
        <p
          className="sc-gsTDqH iORjDB"
        >
          1
        </p>
        <button
          className="sc-dlfmHC jhxSjB"
          onClick={[Function]}
        >
          +
        </button>
      </div>
      <span
        className="sc-iBPTVF jogATO"
      >
        R$ 
        10
      </span>
    </li>
  </ul>
  <div
    className="sc-crrtmM bdeVfT"
  >
    <p
      className="sc-dQpIV kIcqeN"
    >
      Total:
    </p>
    <p
      className="sc-dQpIV kIcqeN"
    >
      R$
      10
    </p>
  </div>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium sc-bqyKbq iEbbsA css-1e6y48t-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    Finalizar Compra
  </button>
</div>
`;
